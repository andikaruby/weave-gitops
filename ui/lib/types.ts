import { Condition, Interval, SourceRefSourceKind } from "./api/core/types.pb";

export enum GrpcErrorCodes {
  Unknown = 2,
  Unauthenticated = 16,
  NotFound = 5,
}

export type RequestError = Error & {
  code?: number;
};

export enum V2Routes {
  Automations = "/automations",
  Sources = "/sources",
  FluxRuntime = "/flux_runtime",
  Kustomization = "/kustomization",
  HelmRelease = "/helm_release",
  HelmRepo = "/helm_repo",
  GitRepo = "/git_repo",

  // Use this to allow for certain components to route to a 404 and still compile.
  // We want to keep certain components around for future use.
  NotImplemented = "/not_implemented",
}

export const WeGONamespace = "flux-system";

export interface Source {
  name?: string;
  namespace?: string;
  type?: SourceRefSourceKind;
  conditions?: Condition[];
  interval?: Interval;
}

export enum AutomationType {
  Kustomization = "Kustomization",
  HelmRelease = "HelmRelease",
}
