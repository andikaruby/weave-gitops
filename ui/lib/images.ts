/* eslint-disable */
// @ts-ignore
import background from "url:../images/background.svg";
// @ts-ignore
import fancyCardBackgroundBlue from "url:../images/fancy-card-background-blue.svg";
// @ts-ignore
import fancyCardBackgroundOrange from "url:../images/fancy-card-background-orange.svg";
// @ts-ignore
import fancyCardBackground from "url:../images/fancy-card-background.svg";
// @ts-ignore
import SignInBackground from "url:../images/SignInBackground.svg";
// @ts-ignore
import SignInWheel from "url:../images/SignInWheel.svg";
// @ts-ignore
import WeaveLogo from "url:../images/WeaveLogo.svg";

export default {
  SignInWheel,
  SignInBackground,
  WeaveLogo,
  background,
  fancyCardBackground,
  fancyCardBackgroundBlue,
  fancyCardBackgroundOrange,
};
