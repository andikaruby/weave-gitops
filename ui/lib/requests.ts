// TokenRefreshWrapper is a singleton service wrapper calls a user-provided
// token refresh function if a 401 is returned
//
// It only does a single refresh at a time, requests that are made during
// the refresh wait for the refresh to finish first.
export class TokenRefreshWrapper {
  private static refreshPromise: Promise<void> | null = null;
  private static refreshTokenFn: () => Promise<void>;

  static wrap(service: any, refreshTokenFn: () => Promise<void>) {
    this.refreshTokenFn = refreshTokenFn;
    return new Proxy(service, {
      get: (target, propKey) => {
        const origMethod = target[propKey];
        if (typeof origMethod === "function") {
          return (...args: any[]) => {
            return this.makeRequest(origMethod.bind(target, ...args));
          };
        }
        return target[propKey];
      },
    });
  }

  private static getOrInitiateRefresh(): Promise<void> {
    if (!this.refreshPromise) {
      this.refreshPromise = this.refreshTokenFn().finally(() => {
        // Set the promise back to null once the refresh operation is completed
        this.refreshPromise = null;
      });
    }
    return this.refreshPromise;
  }

  private static async makeRequest(fn: any): Promise<any> {
    // Wait for any ongoing refresh, if there is one, to complete
    if (this.refreshPromise) {
      await this.refreshPromise;
    }

    try {
      // Call the function directly
      return await fn();
    } catch (error) {
      // Check for a 401 status code on the HTTPError
      if (error.code === 401) {
        await this.getOrInitiateRefresh();
        // Try the request again
        return this.makeRequest(fn);
      }

      // Rethrow the error if it's not about the token
      throw error;
    }
  }
}
