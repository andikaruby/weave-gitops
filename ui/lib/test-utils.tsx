import { MuiThemeProvider } from "@material-ui/core";
import { createMemoryHistory } from "history";
import _ from "lodash";
import * as React from "react";
import { QueryClient, QueryClientProvider } from "react-query";
import { Router } from "react-router-dom";
import { ThemeProvider } from "styled-components";
import AppContextProvider, { AppProps } from "../contexts/AppContext";
import CoreClientContextProvider from "../contexts/CoreClientContext";
import {
  Core,
  GetChildObjectsRequest,
  GetChildObjectsResponse,
  GetReconciledObjectsRequest,
  GetReconciledObjectsResponse,
  GetVersionRequest,
  GetVersionResponse,
} from "./api/core/core.pb";
import theme, { muiTheme } from "./theme";
import { RequestError } from "./types";

export type CoreOverrides = {
  GetChildObjects?: (req: GetChildObjectsRequest) => GetChildObjectsResponse;
  GetReconciledObjects?: (
    req: GetReconciledObjectsRequest
  ) => GetReconciledObjectsResponse;
  GetVersion?: (req: GetVersionRequest) => GetVersionResponse;
};

export const createCoreMockClient = (
  ovr: CoreOverrides,
  error?: RequestError
): typeof Core => {
  const promisified = _.reduce(
    ovr,
    (result, handlerFn, method) => {
      result[method] = (req) => {
        if (error) {
          return new Promise((_, reject) => reject(error));
        }
        return new Promise((accept) => accept(handlerFn(req) as any));
      };

      return result;
    },
    {}
  );

  return promisified as typeof Core;
};

export function withTheme(element) {
  return (
    <MuiThemeProvider theme={muiTheme}>
      <ThemeProvider theme={theme}>{element}</ThemeProvider>
    </MuiThemeProvider>
  );
}

type TestContextProps = AppProps & { api?: Core };

export function withContext(
  TestComponent,
  url: string,
  { api = {}, ...appProps }: TestContextProps
) {
  const history = createMemoryHistory();
  history.push(url);
  const queryClient = new QueryClient();
  const isElement = React.isValidElement(TestComponent);
  return (
    <Router history={history}>
      <AppContextProvider renderFooter {...appProps}>
        <QueryClientProvider client={queryClient}>
          <CoreClientContextProvider api={api as any}>
            {isElement ? TestComponent : <TestComponent />}
          </CoreClientContextProvider>
        </QueryClientProvider>
      </AppContextProvider>
    </Router>
  );
}
