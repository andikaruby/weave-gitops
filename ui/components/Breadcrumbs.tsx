import qs from "query-string";
import React from "react";
import styled from "styled-components";
import useNavigation from "../hooks/navigation";
import { getParentNavValue, V2Routes } from "../lib/nav";
import Flex from "./Flex";
import Icon, { IconType } from "./Icon";
import Link from "./Link";
import Text from "./Text";

const CrumbLink = styled(Link)`
  ${Text} {
    font-size: ${(props) => props.theme.fontSizes.large};
    color: ${(props) => props.theme.colors.neutral00};
  }
`;

export const Breadcrumbs = () => {
  const { currentPage } = useNavigation();
  const parentValue = getParentNavValue(currentPage);

  function pageLookup(route) {
    const parsed = qs.parse(location.search);
    if (route === V2Routes.Automations) return "Applications";
    else if (route === V2Routes.Sources) return "Sources";
    else if (route === V2Routes.FluxRuntime) return "Flux Runtime";
    else return parsed.name;
  }

  return (
    <Flex align>
      {parentValue !== currentPage && (
        <CrumbLink
          to={V2Routes[parentValue as V2Routes] || ""}
          textProps={{ bold: true }}
        >
          {pageLookup(parentValue)}
        </CrumbLink>
      )}
      {parentValue !== currentPage && (
        <Icon type={IconType.NavigateNextIcon} size="large" color="neutral00" />
      )}
      <CrumbLink
        to={currentPage}
        textProps={parentValue === currentPage && { bold: true }}
      >
        {pageLookup(currentPage)}
      </CrumbLink>
    </Flex>
  );
};

export default styled(Breadcrumbs)``;
