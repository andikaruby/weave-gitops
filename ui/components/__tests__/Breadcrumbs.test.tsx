import React from "react";
import renderer from "react-test-renderer";
import { createMockClient, withContext, withTheme } from "../../lib/test-utils";
import Breadcrumbs from "../Breadcrumbs";

describe("snapshots", () => {
  it("renders", () => {
    const tree = renderer
      .create(
        withTheme(
          withContext(<Breadcrumbs />, "/automations", {
            applicationsClient: createMockClient({}),
          })
        )
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
