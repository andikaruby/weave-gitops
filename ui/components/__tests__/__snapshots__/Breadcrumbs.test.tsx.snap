// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshots renders 1`] = `
<div
  className="sc-bdvvtL lacWsj"
>
  <a
    className="sc-hKwDye sc-eCImPb kCA-dBN cQVkzF"
    href="/automations"
    onClick={[Function]}
  >
    <span
      className="sc-gsDKAQ klFDZy"
      color="primary"
      size="normal"
    >
      Applications
    </span>
  </a>
</div>
`;
